### PR TITLE
refactor: migrate detail-view to React (#ACC-239)

### PR DESCRIPTION
**Pull request by przemvs**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/ACC-239" title="ACC-239" target="_blank">TaskACC-239</a>  [Web] Migrate detail-view component to React
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

_Briefly describe the issue you have solved or implemented with this pull request. If the PR contains multiple issues, use a bullet list._

### Causes (Optional)

_Briefly describe the causes behind the issues. This could be helpful to understand the adopted solutions behind some nasty bugs or complex issues._

### Solutions

_Briefly describe the solutions you have implemented for the issues explained above._

### Dependencies (Optional)

_If there are some other pull requests related to this one (e.g. new releases of frameworks), specify them here._

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: src/script/components/Modals/DetailViewModal/DetailViewModalFooter.tsx [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  onDownloadClick: (message: ContentMessage) => void;
+}
+
+const DetailViewModalFooter: FC<DetailViewModalFooterProps> = ({
+  messageEntity,
+  conversationEntity,
+  onLikeClick,
+  onReplyClick,
+  onDownloadClick,
+}) => {
```
Comment (aweiss-dev) [inline src/script/components/Modals/DetailViewModal/DetailViewModalFooter.tsx:43]: Not a fan of using FC.
To be future-proof, I would recommend just typing the props and getting rid of FC.

File: src/script/components/Modals/DetailViewModal/DetailViewModalHeader.tsx [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import {formatLocale} from 'Util/TimeUtil';
+
+import {ContentMessage} from '../../../entity/message/ContentMessage';
+
+interface DetailViewModalHeaderProps {
+  messageEntity: ContentMessage;
+  onCloseClick: () => void;
+}
+
+const DetailViewModalHeader: FC<DetailViewModalHeaderProps> = ({messageEntity, onCloseClick}) => {
```
Comment (aweiss-dev) [inline src/script/components/Modals/DetailViewModal/DetailViewModalHeader.tsx:35]: FC

File: src/script/components/Modals/DetailViewModal/index.tsx [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+
+interface DetailViewModalProps {
+  readonly assetRepository: AssetRepository;
+  readonly conversationRepository: ConversationRepository;
+  readonly messageRepository: MessageRepository;
+  currentMessageEntity: ContentMessage;
+  onClose?: () => void;
+}
+
+const DetailViewModal: FC<DetailViewModalProps> = ({
```
Comment (aweiss-dev) [inline src/script/components/Modals/DetailViewModal/index.tsx:50]: FC

File: src/script/page/MainContent/panels/Collection/Collection.test.tsx [hunk 4]
```diff
@@ -85,10 +89,18 @@ describe('Collection', () => {
   const mockConversationRepository = {
     getEventsForCategory: jest.fn().mockResolvedValue(messages),
     searchInConversation: jest.fn().mockResolvedValue({messageEntities: [createLinkMessage()], query: 'term'}),
-  };
+  } as Partial<ConversationRepository>;
```
Comment (aweiss-dev) [inline src/script/page/MainContent/panels/Collection/Collection.test.tsx:92]: const mockConversationRepository: Partial<ConversationRepository> = {}
 
 would be the preferable way of typing here.

File: src/script/page/MainContent/panels/Collection/CollectionDetails.tsx [hunk 5]
```diff
@@ -55,7 +56,12 @@ const groupByDate = (messages: ContentMessage[]): GroupedCollection => {
   );
 };
 
-const CollectionDetails: React.FC<CollectionDetailsProps> = ({conversation, messages, onClose = noop}) => {
+const CollectionDetails: React.FC<CollectionDetailsProps> = ({
```
Comment (aweiss-dev) [inline src/script/page/MainContent/panels/Collection/CollectionDetails.tsx:59]: FC

File: src/script/page/MainContent/panels/Collection/CollectionItem.tsx [hunk 6]
```diff
@@ -17,43 +17,49 @@
...
 import {isOfCategory} from './utils';
 
-const CollectionItem: React.FC<{allMessages: ContentMessage[]; message: ContentMessage}> = ({message, allMessages}) => {
+interface CollectionItemProps {
+  allMessages: ContentMessage[];
+  message: ContentMessage;
+  onImageClick?: (message: ContentMessage) => void;
+}
+
+const CollectionItem: FC<CollectionItemProps> = ({message, onImageClick}) => {
```
Comment (aweiss-dev) [inline src/script/page/MainContent/panels/Collection/CollectionItem.tsx:36]: FC

File: src/script/view_model/content/MessageListViewModel.ts [hunk 7]
```diff
@@ -202,16 +204,12 @@ export class MessageListViewModel {
...
-    const messageEntities = items.filter(
-      item => item.category & MessageCategory.IMAGE && !(item.category & MessageCategory.GIF),
-    );
-    const [imageMessageEntity] = messageEntities.filter(item => item.id === messageEntity.id);
-
-    amplify.publish(WebAppEvents.CONVERSATION.DETAIL_VIEW.SHOW, imageMessageEntity || messageEntity, messageEntities);
+    showDetailViewModal({
+      assetRepository: this.assetRepository,
+      conversationRepository: this.conversationRepository,
+      currentMessageEntity: messageEntity as ContentMessage,
```
Comment (aweiss-dev) [inline src/script/view_model/content/MessageListViewModel.ts]: Is this as really needed or could it be resolved by stronger typing?

I really don't like the forced deactivation of the ts compiler. I stumbled on a lot of problems already caused by type assertions.

File: src/script/components/Modals/DetailViewModal/index.tsx [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+
+  const getAllImages = async (conversation: Conversation) => {
+    const conversationItems: Message[] = await conversationRepository.getEventsForCategory(
+      conversation,
+      MessageCategory.IMAGE,
+    );
+
+    const messageEntities = conversationItems.filter(item => isOfCategory('images', item as ContentMessage));
+
+    setItems(messageEntities as ContentMessage[]);
```
Comment (aweiss-dev) [inline src/script/components/Modals/DetailViewModal/index.tsx]: Better to use a typeguard

**Review by aweiss-dev — CHANGES_REQUESTED** [review]

File: src/script/components/Modals/DetailViewModal/DetailViewModalFooter.tsx (hunk 1 above)
Comment (przemvs) [reply src/script/components/Modals/DetailViewModal/DetailViewModalFooter.tsx:43]: You mean:
({
```suggestion
const DetailViewModalFooter: FC<DetailViewModalFooterProps> = ({
  messageEntity,
  conversationEntity,
  onLikeClick,
  onReplyClick,
  onDownloadClick,
}: {
messageEntity: MessageEntity,
...rest
})) => {
```

or 

```suggestion
const DetailViewModalFooter = ({
  messageEntity,
  conversationEntity,
  onLikeClick,
  onReplyClick,
  onDownloadClick,
}): DetailViewModalFooterProps => {
```

Comment (przemvs) [reply src/script/components/Modals/DetailViewModal/DetailViewModalFooter.tsx:43]: And why you recommend to not use FC ?

File: src/script/page/MainContent/panels/Collection/Collection.test.tsx (hunk 4 above)
Comment (przemvs) [reply src/script/page/MainContent/panels/Collection/Collection.test.tsx:92]: done.

File: src/script/view_model/content/MessageListViewModel.ts (hunk 7 above)
Comment (przemvs) [reply src/script/view_model/content/MessageListViewModel.ts]: done.

File: src/script/components/Modals/DetailViewModal/index.tsx (hunk 8 above)
Comment (przemvs) [reply src/script/components/Modals/DetailViewModal/index.tsx]: done.

File: src/script/components/Modals/DetailViewModal/DetailViewModalFooter.tsx [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+  const {is_liked: isLiked} = useKoSubscribableChildren(messageEntity, ['is_liked']);
+  const {removed_from_conversation: isRemovedFromConversation} = useKoSubscribableChildren(conversationEntity, [
+    'removed_from_conversation',
+  ]);
+
+  return (
+    <footer className="detail-view-footer">
+      {messageEntity.isReactable() && !isRemovedFromConversation && (
```
Comment (PatrykBuniX) [inline src/script/components/Modals/DetailViewModal/DetailViewModalFooter.tsx:51]: can't it be destructured from `useKoSubscribableChildren` (line 44)?

File: src/script/components/Modals/DetailViewModal/DetailViewModalFooter.tsx [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+          className="detail-view-action-button"
+          onClick={() => onLikeClick(conversationEntity, messageEntity)}
+          data-uie-name="do-like-fullscreen-picture"
+        >
+          <span className={isLiked ? 'icon-liked text-red' : 'icon-like'}></span>
+          <span>{t('conversationContextMenuLike')}</span>
+        </button>
+      )}
+
+      {messageEntity.isReplyable() && !isRemovedFromConversation && (
```
Comment (PatrykBuniX) [inline src/script/components/Modals/DetailViewModal/DetailViewModalFooter.tsx:63]: can't it be destructured from useKoSubscribableChildren (line 44)?

File: src/script/components/Modals/DetailViewModal/DetailViewModalFooter.tsx [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+          onClick={() => onReplyClick(conversationEntity, messageEntity)}
+          data-uie-name="do-reply-fullscreen-picture"
+        >
+          <Icon.Reply />
+
+          <span>{t('conversationContextMenuReply')}</span>
+        </button>
+      )}
+
+      {messageEntity.isDownloadable() && (
```
Comment (PatrykBuniX) [inline src/script/components/Modals/DetailViewModal/DetailViewModalFooter.tsx:76]: can't it be destructured from useKoSubscribableChildren (line 44)?

File: src/script/components/Modals/DetailViewModal/DetailViewModalHeader.tsx [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+  return (
+    <header className="detail-view-header">
+      <div className="text-center">
+        <div
+          className="label-bold-xs"
+          data-uie-name="fullscreen-picture-sender"
+          data-uie-uid={user.id}
+          data-uie-value={user.username()}
+        >
+          {messageEntity.unsafeSenderName()}
```
Comment (PatrykBuniX) [inline src/script/components/Modals/DetailViewModal/DetailViewModalHeader.tsx]: get `unsafeSenderName` from `useKoSubscribableChildren`

File: src/script/guards/Message.ts [hunk 13]
```diff
@@ -17,12 +17,19 @@
...
 import {MemberMessage} from '../entity/message/MemberMessage';
+import {SystemMessage} from '../entity/message/SystemMessage';
 import {SuperType} from '../message/SuperType';
-import {Draft} from 'Util/DraftStateUtil';
+
+export const isContentMessage = (
+  message: Message | ContentMessage | MemberMessage | SystemMessage,
+): message is ContentMessage => message.super_type === SuperType.CONTENT;
+
+export const isDraftMessageWithReplyId = (message: Draft): message is Draft => 'messageId' in message.reply;
```
Comment (PatrykBuniX) [inline src/script/guards/Message.ts]: explicit return type doesn't make sense here imo: you already require to pass `Draft` type, so it's always `Draft`, you are just checking if it has messageId field. I would just remove the return type here 🤔

**Review by aweiss-dev — APPROVED** [review]

File: src/script/components/Modals/DetailViewModal/DetailViewModalFooter.tsx (hunk 9 above)
Comment (przemvs) [reply src/script/components/Modals/DetailViewModal/DetailViewModalFooter.tsx:51]: ### 
image

File: src/script/components/Modals/DetailViewModal/DetailViewModalFooter.tsx (hunk 10 above)
Comment (przemvs) [reply src/script/components/Modals/DetailViewModal/DetailViewModalFooter.tsx:63]: Same like as isReactable.

File: src/script/components/Modals/DetailViewModal/DetailViewModalFooter.tsx (hunk 11 above)
Comment (przemvs) [reply src/script/components/Modals/DetailViewModal/DetailViewModalFooter.tsx:76]: Same like as isReactable.

File: src/script/guards/Message.ts (hunk 13 above)
Comment (przemvs) [reply src/script/guards/Message.ts]: done.

File: src/script/components/Modals/DetailViewModal/DetailViewModalHeader.tsx (hunk 12 above)
Comment (przemvs) [reply src/script/components/Modals/DetailViewModal/DetailViewModalHeader.tsx]: done.

**Review by PatrykBuniX — APPROVED** [review]

**Review by V-Gira — APPROVED** [review]

**Review by PatrykBuniX — APPROVED** [review]

**Pull request merged**